src/neo/SmartContract/ApplicationEngine.cs

```diff
@@ -55,12 +62,39 @@ internal bool AddGas(long gas)
             return testMode || GasConsumed <= gas_amount;
         }
 
+        internal void CallFromNativeContract(Action onComplete, UInt160 hash, string method, params StackItem[] args)
+        {
+            invocationStates.Add(CurrentContext, new InvocationState
+            {
+                ReturnType = typeof(void),
+                Callback = onComplete
+            });
+            CallContract(hash, method, new VMArray(args));
+        }
+
+        internal void CallFromNativeContract<T>(Action<T> onComplete, UInt160 hash, string method, params StackItem[] args)
+        {
+            invocationStates.Add(CurrentContext, new InvocationState
+            {
+                ReturnType = typeof(T),
+                Callback = onComplete
+            });
+            CallContract(hash, method, new VMArray(args));
+        }
+
         protected override void ContextUnloaded(ExecutionContext context)
         {
             base.ContextUnloaded(context);
             if (CurrentContext != null && context.EvaluationStack != CurrentContext.EvaluationStack)
                 if (context.EvaluationStack.Count == 0)
                     Push(StackItem.Null);
+            if (!(UncaughtException is null)) return;
+            if (invocationStates.Count == 0) return;
+            if (!invocationStates.Remove(CurrentContext, out InvocationState state)) return;
+            if (state.Callback is Action action)
+                action();
```

Comment:
```suggestion
                action(engine);
```
As we are calling native contract method, which need to access the engine. Or we don't need `action`, since we can invoke `CallFromNativeContract` more than once.

Reply:
If you use lambda expression, you can access engine directly.